File: java/src/main/java/org/hyperledger/fabric/client/SigningIdentity.java
/*
 * Copyright 2020 IBM All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package org.hyperledger.fabric.client;

import java.security.GeneralSecurityException;
import java.security.ProviderException;
import java.util.function.Function;
import org.hyperledger.fabric.client.identity.Identity;
import org.hyperledger.fabric.client.identity.Signer;
import org.hyperledger.fabric.protos.msp.SerializedIdentity;

final class SigningIdentity {
    private final Identity identity;
    private final Function<byte[], byte[]> hash;
    private final Signer signer;
    private final SerializedIdentity creator;

    SigningIdentity(final Identity identity, final Function<byte[], byte[]> hash, final Signer signer) {
        this.identity = identity;
        this.hash = hash;
        this.signer = signer;

        GatewayUtils.requireNonNullArgument(this.identity, "No identity supplied");
        GatewayUtils.requireNonNullArgument(this.hash, "No hash implementation supplied");
        GatewayUtils.requireNonNullArgument(this.signer, "No signing implementation supplied");

        this.creator = GatewayUtils.newSerializedIdentity(identity);
    }

    Identity getIdentity() {
        return identity;
    }

    byte[] hash(final byte[] message) {
        return hash.apply(message);
    }

    byte[] sign(final byte[] digest) {
        try {
            return signer.sign(digest);
        } catch (GeneralSecurityException e) {
            throw new ProviderException(e);
        }
    }

    byte[] getCreator() {
        return creator.toByteArray();
    }
}

File: web/src/signingidentity.ts
/*
 * Copyright 2020 IBM All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { SerializedIdentity } from '@hyperledger/fabric-protos/lib/msp/identities_pb';
import { ConnectOptions } from './gateway';
import { Identity } from './identity/identity';
import { Signer } from './identity/signer';

type SigningIdentityOptions = Pick<ConnectOptions, 'identity' | 'signer'>;

export class SigningIdentity {
    readonly #identity: Identity;
    readonly #creator: Uint8Array;
    readonly #sign: Signer;

    constructor(options: Readonly<SigningIdentityOptions>) {
        this.#identity = {
            mspId: options.identity.mspId,
            credentials: Uint8Array.from(options.identity.credentials),
        };

        const serializedIdentity = new SerializedIdentity();
        serializedIdentity.setMspid(options.identity.mspId);
        serializedIdentity.setIdBytes(options.identity.credentials);
        this.#creator = serializedIdentity.serializeBinary();

        this.#sign = options.signer;
    }

    getIdentity(): Identity {
        return {
            mspId: this.#identity.mspId,
            credentials: Uint8Array.from(this.#identity.credentials),
        };
    }

    getCreator(): Uint8Array {
        return Uint8Array.from(this.#creator);
    }

    sign(message: Uint8Array): Promise<Uint8Array> {
        return this.#sign(message);
    }

    async hash(message: Uint8Array): Promise<Uint8Array> {
        const hashBuffer = await crypto.subtle.digest('SHA-256', message);
        return new Uint8Array(hashBuffer);
    }
}
